design/bgp-bfd.md

```diff
@@ -0,0 +1,164 @@
+
+When setting a BGP peer, an optional bfd-profile property will enable BFD:
+
+```yaml
+    peers:
+    - peer-address: 10.0.0.1
+      peer-asn: 64501
+      my-asn: 64500
+      bfd-profile: bfdprofile1
+```
+
+#### Impacts on the operator
+
+Keeping the operator as the guinea pig for the CRD implementation, a new BFDProfile CRD will be introduced, with the form of:
+
+```yaml
+apiVersion: metallb.io/v1alpha1
+kind: BFDProfile
+metadata:
+  name: profile
+  namespace: metallb-system
+spec:
+  receive-interval: 150
+  transmit-interval: 150
+  detect-multiplier: 10
+```
+
+Similarly, the BGPPeer CRD that is getting introduced will be configured with a new optional `bfdProfile` field.
+
+Ideally, if/when the CRDs are moved back to MetalLB, it will be possible to enrich the BGPPeer CRD status with information containing the status of the BFD session.
+
+#### Metrics
+
+Metrics describing the status (and the health) of the bfd session between two peers will be produced.
+FRR provides indication of the status of a given session in the form of
+
+```bash
+frr# show bfd peers
+BFD Peers:
+        peer 192.168.0.1
+                ID: 1
+                Remote ID: 1
+                Status: up
+                Uptime: 1 minute(s), 51 second(s)
+                Diagnostics: ok
+                Remote diagnostics: ok
+                Peer Type: dynamic
+                Local timers:
+                        Detect-multiplier: 3
+                        Receive interval: 300ms
+                        Transmission interval: 300ms
+                        Echo receive interval: 50ms
+                        Echo transmission interval: disabled
+                Remote timers:
+                        Detect-multiplier: 3
+                        Receive interval: 300ms
+                        Transmission interval: 300ms
+                        Echo receive interval: 50ms
+```
+
+but also provide indicators on the health of a given session:
+
+```bash
+frr# show bfd peer 192.168.0.1 counters
+     peer 192.168.0.1
+             Control packet input: 126 packets
+             Control packet output: 247 packets
+             Echo packet input: 2409 packets
+             Echo packet output: 2410 packets
+             Session up events: 1
+             Session down events: 0
+             Zebra notifications: 4
+```
+
+### Test Plan
+
+**e2e tests**: E2E tests will be expanded to cover bfd, using external container(s)
+running FRR. The test will need to cover the cases where a node is dropped, verifying
+that the broken route is detected by BFD.
+**unit tests**: Unit tests will be added to any additional code in the MetalLB
+repository.
+
+## Drawbacks
+
+* This design would couple MetalLB with the FRR community, making MetalLB somewhat
+  dependent on the FRR community for bug fixes and further feature improvements.
```

Comment:
I don't really see this as a drawback of this proposal specifically.  It's a drawback of the FRR proposal, and this just builds on top of it.

Reply:
Lol I think this is a leftover from the other proposal I started from